### PR TITLE
Add a way to asynchronously create containers

### PR DESCRIPTION
**Pull request by denkv**

### CODE REVIEW

File: src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java [hunk 1]
```diff
@@ -314,23 +334,6 @@ protected void stopContainer(String containerName) {
...
-    private void initResponseQueue() throws IOException {
-        if (responseQueueName == null) {
-            responseQueueName = cmdChannel.queueDeclare().getQueue();
-        }
-        if (responseConsumer == null) {
-            responseConsumer = new QueueingConsumer(cmdChannel);
-            cmdChannel.basicConsume(responseQueueName, responseConsumer);
-        }
-    }
-
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java]: Why are you not reusing the existing queue?

File: src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java [hunk 2]
```diff
@@ -279,18 +280,37 @@ protected String createContainer(String imageName, String[] envVariables) {
...
+    }
+
+    protected Future<String> createContainerAsync(String imageName, String containerType, String[] envVariables) {
+        try {
+            SettableFuture<String> containerFuture = SettableFuture.create();
+
             envVariables = extendContainerEnvVariables(envVariables);
 
-            initResponseQueue();
+            String responseQueueName = cmdChannel.queueDeclare().getQueue();
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java]: Creating a new queue for each request could make the container creation more expensive :disappointed:

**Review by MichaelRoeder — CHANGES_REQUESTED** [review]
It looks good but I think the response queue should be reused.

**Comment by denkv** [comment]
Reworked to use the single queue for responses again.

File: src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java [hunk 3]
```diff
@@ -325,8 +392,27 @@ private void initResponseQueue() throws IOException {
         if (responseQueueName == null) {
             responseQueueName = cmdChannel.queueDeclare().getQueue();
         }
+        if (responseFutures == null) {
+            responseFutures = new HashMap<>();
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java]: Please create it directly at the beginning (where it is defined).

File: src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java [hunk 4]
```diff
@@ -194,6 +214,11 @@ protected void addCommandHeaderId(String sessionId) {
         acceptedCmdHeaderIds.add(sessionId);
     }
 
+    protected void handleCmd(byte bytes[], AMQP.BasicProperties props) {
+        handleCmd(bytes, props.getReplyTo());
+    }
+
+    @Deprecated
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java]: Please remove `@Deprecated` and add javadoc comments instead.

File: src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java [hunk 5]
```diff
@@ -325,8 +392,27 @@ private void initResponseQueue() throws IOException {
...
+        if (responseFutures == null) {
+            responseFutures = new HashMap<>();
+        }
         if (responseConsumer == null) {
-            responseConsumer = new QueueingConsumer(cmdChannel);
+            responseConsumer = new DefaultConsumer(cmdChannel) {
+                @Override
+                public void handleDelivery(String consumerTag, Envelope envelope, AMQP.BasicProperties properties,
+                        byte[] body) throws IOException {
+                    String key = properties.getCorrelationId();
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java:432]: Check if the correlation ID is set (should always be set but you never know :P)

File: src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java [hunk 6]
```diff
@@ -105,15 +118,22 @@ public void init() throws Exception {
         cmdChannel.exchangeDeclare(Constants.HOBBIT_COMMAND_EXCHANGE_NAME, "fanout", false, true, null);
         cmdChannel.queueBind(queueName, Constants.HOBBIT_COMMAND_EXCHANGE_NAME, "");
 
+        cmdThreadPool = Executors.newCachedThreadPool();
```
Comment (MichaelRoeder) [inline src/main/java/org/hobbit/core/components/AbstractCommandReceivingComponent.java]: 1. Add two constructors. `AbstractCommandReceivingComponent(boolean execCmdsInParallel)` and `AbstractCommandReceivingComponent() {this(false);}`
2. based on the flag, instantiate cmdThreadPool as cached thread pool or single thread pool.

**Comment by denkv** [comment]
Added a workaround for working with PlatformController which does not use correlation IDs in responses.

**Review by MichaelRoeder — APPROVED** [review]
LGTM

**Pull request closed without merging**